### PR TITLE
[MUSIC][GUI]Fix flickering when updating multiple song items via json…

### PR DESCRIPTION
**Pull request by the-black-eagle**
…-rpc

## Description
In advance of tidying up https://github.com/xbmc/xbmc/pull/22654 and dropping the scraper commit, I wrote a python scraper to do the same thing.  Unfortunately, I found that updating lots of songs via json-rpc causes the GUI to flash.

## Motivation and context
This PR fixes the flashing issue on all screens apart from the home screen, where updates are triggered from the music library itself. This will be addressed in a follow-up PR.

The action sequence is as follows, `CAudioLibrary::SetSongDetails()` calls `CMusicDatabase::UpdateSong()` that in turn makes an announcement that results in the home window reloading all  thumbs.  `CAudioLibrary::SetSongDetails()` also calls `NotifyItemUpdated()` which results in a `GUI_MSG_UPDATE` to all windows.  This PR instead sends a `GUI_MSG_UPDATE_ITEM` .  This allows songs to be updated without causing the flashing in the GUI.

## How has this been tested?
This won't be immediately testable as it requires some method of calling into the jsonrpc to update several songs.  I tested it using a version of Kodi built with the changes to the music library and db to store video links.  It could also be tested by updating song comments.

I do have video of updating songs with and without this PR showing the effect clearly.

https://github.com/xbmc/xbmc/assets/18698554/3af2b542-ffb2-45bd-911f-03c090a4f306

https://github.com/xbmc/xbmc/assets/18698554/4f6fd059-3d38-4b4e-90d7-a26e1fbaa175

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/interfaces/json-rpc/JSONRPC.cpp [hunk 1]
```diff
@@ -379,6 +379,15 @@ void CJSONRPCUtils::NotifyItemUpdated()
   CServiceBroker::GetGUI()->GetWindowManager().SendThreadMessage(message);
 }
 
+void CJSONRPCUtils::NotifyItemUpdated(const CSong& song)
+{
+  const CFileItemPtr msgItem(new CFileItem(song));
```
Comment (ksooo) [inline xbmc/interfaces/json-rpc/JSONRPC.cpp:384]:
```c++
const auto msgItem = std::make_shared<CFileItem>(song);
```

File: xbmc/interfaces/json-rpc/JSONRPC.cpp [hunk 2]
```diff
@@ -379,6 +379,15 @@ void CJSONRPCUtils::NotifyItemUpdated()
...
+  CGUIMessage message(GUI_MSG_NOTIFY_ALL,
+                      CServiceBroker::GetGUI()->GetWindowManager().GetActiveWindow(), 0,
+                      GUI_MSG_UPDATE_ITEM, 0, msgItem);
+  CServiceBroker::GetGUI()->GetWindowManager().SendThreadMessage(message);
```
Comment (ksooo) [inline xbmc/interfaces/json-rpc/JSONRPC.cpp:388]: Two calls to  CServiceBroker::GetGUI()->GetWindowManager() in this code block. I would introduce a variable to only have to do this call one time here.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 3]
```diff
@@ -1022,7 +1022,7 @@ JSONRPC_STATUS CAudioLibrary::SetSongDetails(const std::string &method, ITranspo
   if (!musicdatabase.UpdateSong(song, updateartists))
     return InternalError;
 
-  CJSONRPCUtils::NotifyItemUpdated();
+  CJSONRPCUtils::NotifyItemUpdated(song);
```
Comment (ksooo) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:1025]: I would introduce a method taking a const CFileItem& - so you get a generic function that can be used to notify whatever a file item can carry.

**Comment by the-black-eagle** [comment]
@ksooo Thanks for your input.  I've made the required changes (I think) so if you could review again once you have time, I'd appreciate it.  Cheers.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 4]
```diff
@@ -1022,7 +1022,8 @@ JSONRPC_STATUS CAudioLibrary::SetSongDetails(const std::string &method, ITranspo
   if (!musicdatabase.UpdateSong(song, updateartists))
     return InternalError;
 
-  CJSONRPCUtils::NotifyItemUpdated();
+  const auto pItem = std::make_shared<CFileItem>(song);
```
Comment (ksooo) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:1025]: Please do not use type prefix for variable's names. Coding guidelines. https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#88-variables

File: xbmc/interfaces/json-rpc/JSONRPCUtils.h [hunk 5]
```diff
@@ -8,10 +8,12 @@
 
 #pragma once
 
+#include "FileItem.h"
```
Comment (enen92) [inline xbmc/interfaces/json-rpc/JSONRPCUtils.h:11]: Needed? Probably you can just forward declare CFileItem

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp (hunk 4 above)
Comment (the-black-eagle) [reply xbmc/interfaces/json-rpc/AudioLibrary.cpp:1025]: Done

File: xbmc/interfaces/json-rpc/JSONRPCUtils.h (hunk 5 above)
Comment (the-black-eagle) [reply xbmc/interfaces/json-rpc/JSONRPCUtils.h]: Done

**Comment by ksooo** [comment]
jenkins build this please

**Review by enen92 — APPROVED** [review]
Code looks straightforward. +1

**Review by ksooo — APPROVED** [review]
LGTM

**Review by DaveTBlake — CHANGES_REQUESTED** [review]
Since you are changing/fixing the internal functionality that occurs on a JSON API call this needs a version number  patch bump (the last digit), otherwise I'm too out of touch to find anything regarding notification to quibble about :)

**Comment by the-black-eagle** [comment]
@DaveTBlake Thanks, patch version bumped accordingly.

**Review by DaveTBlake — APPROVED** [review]

**Comment by the-black-eagle** [comment]
jenkins build this please

**Pull request merged**